File: app/lib/models/product/product.dart
import 'package:supershop_app/models/product/product_figure.dart';
import 'package:supershop_app/models/product/product_image.dart';
import 'package:supershop_app/models/product/product_variant.dart';
import 'package:supershop_app/models/product/variants_group.dart';

import 'package:json_annotation/json_annotation.dart';

part '../../generated/product/product.g.dart';

@JsonSerializable()
class Product {
  late String id;
  late String shopId;
  late String categoryId;
  late String name;
  late double? price;
  late int? stockQuantity;
  late String description;
  late String conditionProduct;
  late bool isVariant;
  late bool isActive;
  late bool isDeleted;
  late String brand;
  late String status;
  late List<ProductImage> productImages;
  late List<VariantsGroup>? variantsGroup;
  late List<ProductVariant>? productVariants;
  late ProductFigure productFigure;
  late String createdAt;
  late String updatedAt;
  Product({
    required this.id,
    required this.shopId,
    required this.categoryId,
    required this.name,
    this.price,
    this.stockQuantity,
    required this.description,
    required this.conditionProduct,
    required this.isVariant,
    required this.isActive,
    required this.isDeleted,
    required this.brand,
    required this.status,
    required this.productImages,
    this.variantsGroup,
    this.productVariants,
    required this.productFigure,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Product.fromJson(Map<String, dynamic> json) =>
      _$ProductFromJson(json);

  Map<String, dynamic> toJson() => _$ProductToJson(this);
}

File: api/src/main/java/com/project/supershop/features/product/domain/entities/Product.java
package com.project.supershop.features.product.domain.entities;

import com.project.supershop.common.BaseEntity;
import com.project.supershop.features.account.domain.entities.Account;
import com.project.supershop.features.product.domain.dto.requests.ProductRequest;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;

import java.util.List;

@Entity
@Table(name = "products")
@NoArgsConstructor
@AllArgsConstructor
@Data
@SuperBuilder
public class Product extends BaseEntity {
    private String categoryId;
    private String name;
    private Double price;
    private Integer stockQuantity;
    @Column(length = 5000)
    private String description;
    private String conditionProduct;
    private String brand;
    private String status;
    private Boolean isVariant;
    private Boolean isActive;
    private Boolean isDeleted;

    @OneToMany(mappedBy = "product")
    @EqualsAndHashCode.Exclude
    private List<ProductImage> productImages;

    @OneToMany(mappedBy = "product")
    @EqualsAndHashCode.Exclude
    private List<VariantGroup> variantsGroup;

    @OneToMany(mappedBy = "product")
    @EqualsAndHashCode.Exclude
    private List<ProductVariant> productVariants;

    @OneToMany(mappedBy = "product")
    @EqualsAndHashCode.Exclude
    private List<HistoryViolation> historyViolations;

    @OneToMany(mappedBy = "product")
    @EqualsAndHashCode.Exclude
    private List<ShopProductsCategory> shopProductsCategories;

    @ManyToOne
    @JoinColumn(name = "shopId")
    private Account shop;

    @OneToOne(mappedBy = "product")
    @EqualsAndHashCode.Exclude
    private ProductFigure productFigure;

    public static Product createProduct(ProductRequest productRequest, String status, Account shop){
        return Product.builder()
                .shop(shop)
                .categoryId(productRequest.getCategoryId())
                .name(productRequest.getName())
                .price(productRequest.getPrice())
                .stockQuantity(productRequest.getStockQuantity())
                .conditionProduct(productRequest.getConditionProduct())
                .description(productRequest.getDescription())
                .brand(productRequest.getBrand())
                .status(status)
                .isVariant(productRequest.getIsVariant())
                .isActive(productRequest.getIsActive())
                .isDeleted(false)
                .build();
    }
}
